### PR TITLE
Faster Faster BatchSampler

### PR DESCRIPTION
**Pull request by xangma**
Builds upon #76951.

Benchmarking code is the same as in #76950.

AMD Ryzen Threadripper PRO 3995WX:
```
  batch_size  drop_last      origin     new  speedup
------------  -----------  --------  ------  ---------
           4  True           0.94    0.5706  64.74%
           4  False          0.9745  0.9468  2.93%
           8  True           0.7423  0.3715  99.82%
           8  False          0.7974  0.5666  40.73%
          64  True           0.5394  0.2085  158.76%
          64  False          0.6083  0.2697  125.51%
         640  True           0.5448  0.1985  174.41%
         640  False          0.7085  0.2308  206.91%
        6400  True           0.5554  0.2028  173.88%
        6400  False          0.7711  0.2109  265.60%
       64000  True           0.556   0.2091  165.82%
       64000  False          0.7803  0.2078  275.58%
```

When `drop_last == True`, it uses `zip` to speed things up.
When `drop_last == False`, it uses `itertools` to speed things up.

`itertools` was the fastest way I could find that deals with the last batch if it is smaller than `batch_size`. I have a pure python method too, but it is slower when `batch_size` is 4 or 8, so I have committed the `itertools` version for now. 

Happy to chat further about this change :-) I understand you may not want to introduce the `itertools` package into [sampler.py](https://github.com/pytorch/pytorch/blob/main/torch/utils/data/sampler.py).

### CODE REVIEW

**Comment by divyanshk** [comment]
Thanks for the PR, super curious how the results stand for even smaller batch sizes.

(We probably should have a uniform script to benchmark and check correctness, I can maybe look into that!).

File: torch/utils/data/sampler.py [hunk 1]
```diff
@@ -334,25 +335,18 @@ def __init__(
     def __iter__(self) -> Iterator[List[int]]:
         # Implemented based on the benchmarking in https://github.com/pytorch/pytorch/pull/76951
         if self.drop_last:
+            sampler_iter = iter(self.sampler)
+            # Create multiple references to the same iterator
+            args = [sampler_iter] * self.batch_size
```
Comment (andrewkho) [inline torch/utils/data/sampler.py:340]: It's unclear to me that this is safe in all scenarios although on the surface it seems OK. What benefit does this give us though?

Comment (xangma) [reply torch/utils/data/sampler.py:340]: It's my understanding that it creates the required references first and then passes them all to `zip`, meaning it doesn't create a reference for each loop? I found [this SO post](https://stackoverflow.com/questions/2233204/how-does-zipitersn-work-in-python#:~:text=iter()%20is%20an%20iterator%20over%20a) quite informative on it.

I don't have the level of understanding of the pytorch codebase as either of you will, but I'm happy to work with you to test more things if that would be helpful. I found that the original list comprehension method was taking time when looping over large batches of small data when benchmarking some code with a colleague (@SusannaGreen).

**Comment by xangma** [comment]
> Thanks for the PR, super curious how the results stand for even smaller batch sizes.
> 
> (We probably should have a uniform script to benchmark and check correctness, I can maybe look into that!).

I have some (quite janky) batch verification code I used with the aforementioned benchmarking code. I've put it in a gist [here](https://gist.github.com/xangma/5bf5e43253315e3e03f4fd9c935e3205). It's janky because to test the `drop_last` function for the itertools change, you need to manually set `DATA_SIZE` to be smaller than `VERIFICATION_BATCHES * BATCH_SIZE` . I didn't have much time to spend on it sorry 😅 There's an example output as a comment to the gist :-)

**Comment by xangma** [comment]
Fixed for lintrunner requirements. 

New benchmarks:
AMD Ryzen Threadripper PRO 3995WX:
```
  batch_size  drop_last      origin     new  speedup
------------  -----------  --------  ------  ---------
           4  True           0.8761  0.5565  57.44%
           4  False          0.978   1.1248  -13.05%
           8  True           0.7094  0.3651  94.31%
           8  False          0.7892  0.6649  18.69%
          64  True           0.535   0.2006  166.77%
          64  False          0.5954  0.3011  97.77%
         640  True           0.5205  0.1853  180.88%
         640  False          0.6293  0.2435  158.43%
        6400  True           0.5219  0.1898  175.03%
        6400  False          0.6602  0.2167  204.61%
       64000  True           0.5214  0.197   164.66%
       64000  False          0.6677  0.2679  149.26%
```

Apple M1 Max:
```
  batch_size  drop_last      origin     new  speedup
------------  -----------  --------  ------  ---------
           4  True           0.6173  0.4025  53.38%
           4  False          0.8404  0.8706  -3.46%
           8  True           0.5214  0.2833  84.04%
           8  False          0.6657  0.5511  20.78%
          64  True           0.358   0.1763  103.01%
          64  False          0.4884  0.2625  86.10%
         640  True           0.3749  0.1581  137.04%
         640  False          0.5404  0.2072  160.80%
        6400  True           0.3879  0.1683  130.45%
        6400  False          0.5752  0.1923  199.05%
       64000  True           0.3884  0.1767  119.83%
       64000  False          0.5906  0.1938  204.68%
```

Now it's slower for `batch_size == 4` Silly linting :'(

**Comment by Skylion007** [comment]
> Fixed for lintrunner requirements.
> 
> New benchmarks: AMD Ryzen Threadripper PRO 3995WX:
> 
> ```
>   batch_size  drop_last      origin     new  speedup
> ------------  -----------  --------  ------  ---------
>            4  True           0.8761  0.5565  57.44%
>            4  False          0.978   1.1248  -13.05%
>            8  True           0.7094  0.3651  94.31%
>            8  False          0.7892  0.6649  18.69%
>           64  True           0.535   0.2006  166.77%
>           64  False          0.5954  0.3011  97.77%
>          640  True           0.5205  0.1853  180.88%
>          640  False          0.6293  0.2435  158.43%
>         6400  True           0.5219  0.1898  175.03%
>         6400  False          0.6602  0.2167  204.61%
>        64000  True           0.5214  0.197   164.66%
>        64000  False          0.6677  0.2679  149.26%
> ```
> 
> Apple M1 Max:
> 
> ```
>   batch_size  drop_last      origin     new  speedup
> ------------  -----------  --------  ------  ---------
>            4  True           0.6173  0.4025  53.38%
>            4  False          0.8404  0.8706  -3.46%
>            8  True           0.5214  0.2833  84.04%
>            8  False          0.6657  0.5511  20.78%
>           64  True           0.358   0.1763  103.01%
>           64  False          0.4884  0.2625  86.10%
>          640  True           0.3749  0.1581  137.04%
>          640  False          0.5404  0.2072  160.80%
>         6400  True           0.3879  0.1683  130.45%
>         6400  False          0.5752  0.1923  199.05%
>        64000  True           0.3884  0.1767  119.83%
>        64000  False          0.5906  0.1938  204.68%
> ```
> 
> Now it's slower for `batch_size == 4` Silly linting :'(

I mean you can ignore the linting with special comments if it affects speed...

**Comment by Skylion007** [comment]
I buy the itertools islice is faster for sure. It's a shame we can't use itertools.batched both because of our minimum Python version and because it returns tuple instead of lists lol.

**Review by Skylion007 — APPROVED** [review]
This looks good to me, and the speed up makes sense me based on how itertools.islice is implemented and how zip is implemented.

File: torch/utils/data/sampler.py [hunk 2]
```diff
@@ -333,26 +334,18 @@ def __init__(
...
+            for batch in zip(*args):
+                yield list(batch)
+        else:
             while True:
-                try:
-                    batch = [next(sampler_iter) for _ in range(self.batch_size)]
-                    yield batch
-                except StopIteration:
+                batch: List[int] = list(itertools.islice(sampler_iter, self.batch_size))  # type: ignore[no-redef]
```
Comment (Skylion007) [inline torch/utils/data/sampler.py:345]: We can just call it something else to remoe the redef typing error? No reason to reuse the name if we don't need to.

**Comment by Skylion007** [comment]
Minus those nits: it looks good to me.

File: torch/utils/data/sampler.py [hunk 3]
```diff
@@ -333,26 +334,18 @@ def __init__(
...
-            for idx in self.sampler:
-                batch[idx_in_batch] = idx
-                idx_in_batch += 1
-                if idx_in_batch == self.batch_size:
-                    yield batch
-                    idx_in_batch = 0
-                    batch = [0] * self.batch_size
-            if idx_in_batch > 0:
-                yield batch[:idx_in_batch]
+                yield list(batch)
```
[the remark this answers is not in the document]
Comment (xangma) [reply torch/utils/data/sampler.py]: I think it was for linting 😬

File: torch/utils/data/sampler.py (hunk 2 above)
Comment (xangma) [reply torch/utils/data/sampler.py]: That makes a lot of sense. Thanks!

**Comment by xangma** [comment]
Annoyingly, for the `itertools` part, if you don't convert to a list before yielding, it's seemingly really slow.

The zip part is slower now, I just need a bit of time to fix that then I think I'm good.

**Comment by xangma** [comment]
Ok. I think I've tidied it up now. 

Here's the benchmarks. Seems to pass the lintrunner fine :-) 

AMD Ryzen Threadripper PRO 3995WX:
```
  batch_size  drop_last      origin     new  speedup
------------  -----------  --------  ------  ---------
           4  True           0.8708  0.4741  83.69%
           4  False          0.9435  0.8576  10.01%
           8  True           0.7023  0.3281  114.02%
           8  False          0.7701  0.507   51.88%
          64  True           0.5286  0.1925  174.60%
          64  False          0.5895  0.2584  128.11%
         640  True           0.5206  0.1856  180.52%
         640  False          0.629   0.2208  184.92%
        6400  True           0.5194  0.1905  172.68%
        6400  False          0.6601  0.1933  241.47%
       64000  True           0.519   0.1965  164.04%
       64000  False          0.6632  0.1883  252.13%
```

Apple M1 Max:
```
  batch_size  drop_last      origin     new  speedup
------------  -----------  --------  ------  ---------
           4  True           0.6297  0.3192  97.27%
           4  False          0.849   0.6475  31.10%
           8  True           0.4987  0.2378  109.75%
           8  False          0.6554  0.4149  57.96%
          64  True           0.3536  0.1767  100.15%
          64  False          0.4896  0.2304  112.52%
         640  True           0.3723  0.1552  139.86%
         640  False          0.5375  0.1879  186.06%
        6400  True           0.3797  0.169   124.62%
        6400  False          0.5773  0.1691  241.34%
       64000  True           0.3815  0.18    111.93%
       64000  False          0.5846  0.1674  249.19%
       ```

**Comment by xangma** [comment]
Hmmm. A bunch of the trunk tests failed after merging. I think it's to do with changes in this PR: #136519 ?

@Skylion007 @divyanshk

**Pull request closed without merging**